llvm/lib/Target/X86/X86InstrInfo.cpp

```diff
@@ -5169,7 +5169,7 @@ bool X86InstrInfo::optimizeCompareInstr(MachineInstr &CmpInstr, Register SrcReg,
     // Fall through to optimize Cmp if Cmp is CMPrr or CMPri.
     if (NewOpcode == X86::CMP64rm || NewOpcode == X86::CMP32rm ||
         NewOpcode == X86::CMP16rm || NewOpcode == X86::CMP8rm)
-      return false;
+      return true;
```

Comment:
Why CMPrm don't need to check EFLAGS before removing?

Reply:
The following is not to check EFLAGS, but for more aggressive optimizations for EFLAGS.  Memory analysis is not safe in MIR pass, so we don't do that for CMPrm.

Reply:
My concern was, if the following code is a further optimization, it should never reture false because we has the base one. Or it is not correct to return true here.

Reply:
Ah, good point. It's still buggy.

Reply:
```
      if ((isUncoalescableCopy(*MI) &&
           optimizeUncoalescableCopy(*MI, LocalMIs)) ||
          (MI->isCompare() && optimizeCmpInstr(*MI)) ||
          (MI->isSelect() && optimizeSelect(*MI, LocalMIs))) {
        // MI is deleted.
        LocalMIs.erase(MI);
        Changed = true;
        continue;
      }

      if (MI->isConditionalBranch() && optimizeCondBranch(*MI)) {
        Changed = true;
        continue;
      }

      if (isCoalescableCopy(*MI) && optimizeCoalescableCopy(*MI)) {
        // MI is just rewritten.
        Changed = true;
        continue;
      }
```
According to the call site，it's wrong whether it returns `true` or `false`.

`true` breaks the assumption that  `optimizeCmpInstr` deletes the MI.
`false` makes the flag `-print-changed` not work correctly.

Reply:
Maybe we should add a parameter `bool &Deleted` for `optimizeCmpInstr` as `foldImmediate`.